scipy/special/sf_error.cc

```diff
@@ -54,10 +62,13 @@ void sf_error_v(const char *func_name, sf_error_t code, const char *fmt, va_list
         PyOS_snprintf(msg, 2048, "scipy.special/%s: %s", func_name, sf_error_messages[(int) code]);
     }
 
+#ifndef Py_GIL_DISABLED
 #ifdef WITH_THREAD
     save = PyGILState_Ensure();
 #endif
-
+#else
+    PyMutex_Lock(&err_mutex);
+#endif
```

Comment:
Can you explain more why you're not calling `PyGILState_Ensure` and `PyGILState_Release` on the GIL-enabled build? If this is some kind of error callback mechanism (which it looks like it is) you would still need to have a valid thread state to be able to call into the C API. If a scipy function spawned a thread (say, via OpenMP, without going through the `threading` module) that called one of these error callbacks before registering the thread with the interpreter with `PyGILState_Ensure`, that would crash. Did you add the ifdefs to avoid a deadlock or something like that?

Reply:
Sorry, I wasn't aware of the purpose of `PyGILState_STATE`, I thought it was related to something about the GIL and therefore it needed culling under free-threaded circumstances. I've just restored them on https://github.com/scipy/scipy/pull/21956/commits/87e5f3bf1da498f7d381e0ee33d805b74ad80361.

The mutex there is to prevent concurrent writing to the module attribute that's being used to keep track of errors.